codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -96,6 +104,71 @@ public DefaultHttp2Connection(boolean server) {
         streamMap.put(connectionStream.id(), connectionStream);
     }
 
+    /**
+     * Determine if {@link #close(Promise)} has been called and no more streams are allowed to be created.
+     */
+    final boolean isClosed() {
+        return closePromise != null;
+    }
+
+    @Override
+    public Future<Void> close(final Promise<Void> promise) {
+        checkNotNull(promise, "promise");
+        // Since we allow this method to be called multiple times, we must make sure that all the promises are notified
+        // when all streams are removed and the close operation completes.
+        if (closePromise != null) {
+            if (closePromise == promise) {
+                // Do nothing
+            } else if ((promise instanceof ChannelPromise) && ((ChannelPromise) closePromise).isVoid()) {
+                closePromise = promise;
+            } else {
+                closePromise.addListener(new FutureListener<Void>() {
+                    @Override
+                    public void operationComplete(Future<Void> future) throws Exception {
+                        if (future.isSuccess()) {
+                            promise.trySuccess(null);
+                        } else if (future.isCancelled()) {
+                            promise.cancel(false);
+                        } else {
+                            promise.tryFailure(future.cause());
+                        }
+                    }
+                });
+            }
+        } else {
+            closePromise = promise;
+        }
+        if (isStreamMapEmpty()) {
+            promise.trySuccess(null);
+            return promise;
```

Comment:
Just curious ... not sure how Netty handles this case in general: what if `promise` is a void promise? If the caller tries to call `addListener` it will fail ... maybe we don't care?